### ISSUE TITLE
Orli doesn't run well but I found a fix.

### ISSUE THREAD

**Issue by kkkamur07** [mittagessen/orli#11]

**Comment by johnlockejrr** [comment]
No sh*t! 🥇

**Comment by kkkamur07** [comment]
Yes! Writing the fix here :) it blew my mind when i compared it with kraken -> solves a shit ton of my problems.

**Comment by kkkamur07** [comment]
Here is the fix that works for me, with the evidence. Thanks for Orli, the results are excellent once it runs.

### Symptom

With the current `main` and the published `orli_base.safetensors` (Zenodo 10.5281/zenodo.20558179, 2026-06-05), segmentation emits 5 to 15 good baselines, then repeats one baseline in the same spot until `max_predicted_lines`. The end token never fires. This is what #10 describes too.

### Cause, as far as I can tell

The published weights and `main` are out of step. `7bf383e` ("fp32 curve predictor + prompt noising", 2026-08-29) runs `CurveRegressionHead` under `autocast(enabled=False)` in fp32 and replaces the dtype-aware clamp in `inverse_sigmoid` (`eps = max(eps, torch.finfo(x.dtype).eps)`, about 0.0078 in bf16) with a fixed fp32 `1e-5`. The June weights were trained with the old bf16 behaviour, so the refinement head sees inputs it was never trained on. This matches the comment in #10 that the retrained weights are not published yet.

### Workaround until the new weights are out

Pin Orli to the commit that matches the published weights, pin kraken to the 7.0 line, and use real bf16:

```bash
git clone https://github.com/mittagessen/orli.git && cd orli
git checkout 7996f29
uv venv --python 3.12
uv pip install "kraken==7.0.3" timm python-bidi rich
uv pip install --no-deps .
```

Then run with precision `bf16-true`. Note that `bf16-mixed` does not help: inference never enters autocast, so the model silently runs in fp32 and loops.

### Evidence

Controlled A/B: same page, same kraken 7.0.3, same `bf16-true`, only the Orli commit changed. `main` runs to the cap. `7996f29` returns 63 lines on a page with 63 ground-truth lines, and 127 on a page with 127.

On 10 Coptic manuscript pages (two manuscripts, two to four columns), against the default kraken blla model:

| Metric | kraken blla | Orli at 7996f29 |
|---|---|---|
| F1 (cBAD-style baseline matching) | 0.727 | 0.852 |
| Precision | 0.632 | 0.849 |
| Fragmented lines | 116 | 6 |
| Spurious lines | 228 | 36 |
| Columns with the exact line count | 0 of 28 | 25 of 28 |
| Reading order correct | | 10 of 10 pages |

What I tested:

| Setup | Result |
|---|---|
| Apple M4, MPS, `bf16-true` | works |
| x86 CPU without bf16 kernels (Skylake), `bf16-true` | works, same line counts, slow |
| fp32 (`32-true`) at 7996f29, MPS and CPU | loops to the cap |

### One more thing that may help others

Two-folio spreads (microfilm images with both pages and a gutter) break the stopping rule even at the pinned commit: one spread gave 26 correct lines followed by about 50 tiny junk lines, another stopped after a single line. Both are deterministic. Cutting the same images into single folios fixes it completely (26, and 27 plus 68 lines, with a table read column by column). Multi-column single leaves were fine unsplit.

Happy to share pages or scripts if useful, and to re-test as soon as the retrained weights are on Zenodo.

(Diagnosis and benchmarks were done with the help of Claude Code.)

**Comment by johnlockejrr** [comment]
Wow! You and Claude reinvented the wheel! Why didn't read the README.md?

<img width="1314" height="611" alt="Image" src="https://github.com/user-attachments/assets/14bdf355-9b8c-428f-ab86-d99e8645f08f" />

@mittagessen will be very pleased with another LLM breakthrough ;-)

**Comment by kkkamur07** [comment]
No! One more error is that this only works with the kraken version pinned to 7.0.3  not the current one. It cascades into loop when we use the latest kraken version. I did read the readme haha :) using emulated bf-16 to run it on CPU haha :)

Also when you do it folio by folio then it helps and when you do it with multiple folio it doesn't help as well. Sorry that I used claude to write this.

**Comment by johnlockejrr** [comment]
I don't have anything against using LLM :)
For me works on latest kraken also. Anyway, this really needs a fix no matter what.

**Comment by kkkamur07** [comment]
I am sorry for using LLM without reviewing the output. Definitely we need @mittagessen to update the weights.

**Comment by mittagessen** [comment]
Hello,

the current state of the repo and the released weights are trained with
regression heads that use model-global precision/datatype-specific
epsilons. That throws the autoregression off anytime you run the base
model in another precision than bf16-mixed but there is also some
runaway generation that seems to be caused by slight numerical
differences between platforms/library versions.

You can conceivably fix the problem on inference side without touching
the weights but this is most likely fairly brittle. I have made some
slight changes to the architecture that should fix the most glaring
issues and fine-tuned the weights for another epoch. It will need some
verification if it actually worked and I'm currently swamped with other
commitments. If nothing else ambushes me, it'll be another 2-3 weeks
before I can run the evaluations to make sure nothing went awry.

**Comment by mittagessen** [comment]
I decided to just put up the weights here:
https://l.unchti.me/orli_fixed.safetensors

You can play around with them but as I said there are no guarantuees for
anything.

BTW, I saw that you're working on a Byzantine/Armenian/Syriac HTR
platform. From what I'm gathering you're probably pretraining on
everything in-domain you could find on the web and then fine-tune on
your published data afterwards? Error rates are usually better when
fine-tuning directly from very general base models. If you feel like it
you could try fine-tuning on top of kraken's adapted PP-OCRv6 base model
here [0]. Also for Greek you should probably switch to Unicode NFD. NFC
tends to produce models that perform a percentage point or two worse
than NFD ones.

[0] https://htrmopo.inria.fr/models/10.5281-zenodo.21788409/pp-ocrv6-medium-multilingual-handwritten-and-printed-text-recognition-model

**Comment by kkkamur07** [comment]
Thanks so much @mittagessen! Fan of you work. I'll definitely play around with the weights. Small side note: I got caching implemented in Orli so it runs faster on CPU (we only have CPU-only inference servers with ~16 GB, haha).

A bit more on the project: nomikos.app is a script-agnostic editor (starting with Greek, Syriac, Coptic and Armenian) built around these models. The goal is to shift a researcher's workflow from doing segmentation and transcription by hand to reviewing and correcting model output ( 10x faster ), with those corrections flowing back into training. 

Models and datasets are at https://huggingface.co/nomikos-project, and the platform is live at nomikos.app. ( I would really love a Github Star from you ). 

Attaching the results of fine-tuning on general models.

Adapted PP-OCRv6 has been excellent for us, but so far only for Syriac (where it's very, very good) it is outperforming our architecture. For Greek, Coptic and Armenian ( it is very competitive ), our extended Calamari architecture ( pre-trained + finetuned ) is outperforming fine-tuned PP-OCRv6. I'd love your perspective on why that might be :)

Noted on NFD for Greek, we'll switch the normalization and rerun the ablations :)

<img width="1096" height="618" alt="Image" src="https://github.com/user-attachments/assets/e6f248da-9252-4846-a199-19508303bb95" />

<img width="1094" height="613" alt="Image" src="https://github.com/user-attachments/assets/83b1133a-b5cf-4f17-91a4-8d7e8d1b37cf" />

<img width="1093" height="616" alt="Image" src="https://github.com/user-attachments/assets/39ab8b10-8734-473c-8477-2aa0aaf4b456" />

**Comment by johnlockejrr** [comment]
@kkkamur07, just curious, for Syriac you work only on Eastern or Serto also? I saw a dataset of yours on HF starting with Eastern data. Courious because is also in my domain of work.

**Comment by mittagessen** [comment]
On 26/09/21 04:38AM, Krrish Agarwalla wrote:
> Adapted PP-OCRv6 has been excellent for us, but so far only for Syriac (where it's very, very good) it is outperforming our architecture. For Greek, Coptic and Armenian ( it is very competitive ), our extended Calamari architecture ( pre-trained + finetuned ) is outperforming fine-tuned PP-OCRv6. I'd love your perspective on why that might be :)

Are you using paddle paddle's PP-OCRv6? They train it weirdly and with
hyperparameters that are less than ideal for historical writing. The out
of the box results were so poor I never investigated much where its
strengths or weaknesses lie. The kraken version performs quite a bit
better but is slightly slower due to the larger input height.

The single manuscript evaluation you're doing is also fairly useless to
estimate real-world HTR accuracy. Good practice is to split
across manuscripts and ideally have a paleographer stratify hand
similarity/difficulty to identify the weak spots of the model. At those
dataset sizes I suspect most of the effect sizes you're claiming are
just artefacts of training initialization and data splits.

**Comment by kkkamur07** [comment]
> Are you using paddle paddle's PP-OCRv6? They train it weirdly and with
hyperparameters that are less than ideal for historical writing. The out
of the box results were so poor I never investigated much where its
strengths or weaknesses lie. The kraken version performs quite a bit
better but is slightly slower due to the larger input height

We used paddles version of PPOCR for segmentation but we used Kraken Adapted PPOCR for Syriac and other scripts and as we have a onnx runtime hosted so we ported torch to onnx. 

> The single manuscript evaluation you're doing is also fairly useless to
estimate real-world HTR accuracy. Good practice is to split
across manuscripts and ideally have a paleographer stratify hand
similarity/difficulty to identify the weak spots of the model. At those
dataset sizes I suspect most of the effect sizes you're claiming are
just artefacts of training initialization and data splits.

I partly agree tbh! You’re right that we need more data for evaluation. But we’ve evaluated on multiple manuscripts, but we can’t release all of them due to copyright restrictions. For example we’ve only released the manuscripts we had permission to share, like Alqosh 65 for Syriac. The same applies to the other scripts as well.

Also models we published have received good feedback from the researchers working with these scripts, they usually test these models on different  manuscripts. 

But that said everything we have done, is built on the strong foundations of kraken and escriptorium thank you for establishing them.

**Comment by kkkamur07** [comment]
> [@kkkamur07](https://github.com/kkkamur07), just curious, for Syriac you work only on Eastern or Serto also? I saw a dataset of yours on HF starting with Eastern data. Courious because is also in my domain of work.

Ayee yes! We are currently students at LMU Munich and here there a group working ( https://www.kw.lmu.de/byz/de/personen/ )  on understanding the legal traditions in medieval period in Syriac, Coptic, Armenian and Byzantine Greek and they have a lot of manuscripts to deciper so they approached us ( the AI lab to build this for them and this was a very interesting project to be honest ). 

TBH, We are very very new to the space and still trying to understand the technicalities. Pardon if we are making any mistakes :)

**Comment by mittagessen** [comment]
On 26/09/21 09:39AM, Krrish Agarwalla wrote:
> I partly agree tbh! You’re right that we need more data for evaluation. But we’ve evaluated on multiple manuscripts, but we can’t release all of them due to copyright restrictions. For example we’ve only released the manuscripts we had permission to share, like Alqosh 65 for Syriac. The same applies to the other scripts as well.

OK, from the quantitative charts you pasted it looked like the data on
huggingface was the only sets you're evaluating on.

**Comment by kkkamur07** [comment]
Ahh sorry! I shall update them then :)

Best,
Krrish
https://kkrrish.tech/

**Comment by johnlockejrr** [comment]
> Ahh sorry! I shall update them then :)
> 
> Best,
> Krrish
> https://kkrrish.tech/
> […](#)

Tried some manuscript and print pages on Syriac on your platform, still much work to do but is a cool project.

**Comment by kkkamur07** [comment]
> Tried some manuscript and print pages on Syriac on your platform, still much work to do but is a cool project.

Aye! I would appreciate feedback or any problems you faced ? Which model did you use and how did it perform ?  Also if you like the project would love a github star from you.

**Comment by johnlockejrr** [comment]
> > Tried some manuscript and print pages on Syriac on your platform, still much work to do but is a cool project.
> 
> Aye! I would appreciate feedback or any problems you faced ? Which model did you use and how did it perform ? Also if you like the project would love a github star from you.

On print very bad. And I know that because of the scarcity of quality datasets for Syriac. I tried calamari you said has a good CER. Can't say that.

<img width="3828" height="1809" alt="Image" src="https://github.com/user-attachments/assets/951fac26-1f22-4b80-b40a-d4bc2b373d6e" />

**Comment by kkkamur07** [comment]
> On print very bad. And I know that because of the scarcity of quality datasets for Syriac. I tried calamari you said has a good CER. Can't say that.

Ohh you should try syriac-ppocr! ( that is our best model, calamari performs very poorly around 20% CER, ppOCR 7% ) I don't know if it will generalize to prints as we primarily trained it for HTR, worth a try :)

**Comment by johnlockejrr** [comment]
> > On print very bad. And I know that because of the scarcity of quality datasets for Syriac. I tried calamari you said has a good CER. Can't say that.
> 
> Ohh you should try syriac-ppocr! ( that is our best model, calamari performs very poorly around 20% CER, ppOCR 7% ) I don't know if it will generalize to prints as we primarily trained it for HTR, worth a try :)

On easy manuscripts is pretty decent the ppocr. Still need more work on segmentation. If you use PP-OCR try `PP-OCRv6_medium_det` is their best for text line detection. Maybe finetune it on manuscripts.

<img width="2622" height="1594" alt="Image" src="https://github.com/user-attachments/assets/f34473bc-3958-4eff-8b9c-a74e3f0b2182" />

**Comment by kkkamur07** [comment]
Lovely! Thank you we will work on segmentation, looking forward to you using this :)

**Comment by johnlockejrr** [comment]
> Lovely! Thank you we will work on segmentation, looking forward to you using this :)

Thanks! Just checking out I mostly use eScriptorium but my own project lately also. Maybe you move this discussion to your project? Not beeing orli-related so we don't spam @mittagessen 

<img width="3796" height="2021" alt="Image" src="https://github.com/user-attachments/assets/f9f74f20-1dcf-4209-b285-d9870750d557" />

**Comment by johnlockejrr** [comment]
By the way, you could simplify the polygons, sometimes they are a soup of hundreds 💃 

<img width="992" height="510" alt="Image" src="https://github.com/user-attachments/assets/8bec0a88-4471-490b-9843-c06b8d911b17" />

**Comment by kkkamur07** [comment]
@mittagessen thank you for orli weights, I will try to play around with them.
